### ISSUE TITLE
[BUG]loss error when using MLA

### ISSUE THREAD

**Issue by Desperadoze** [NVIDIA/Megatron-LM#1445]
**Describe the bug**
when using mla by megatron-core 0.10.0,loss was unexpected

**To Reproduce**
use  /nvcr.io/nvidia/nemo:25.02.rc1   image start traing

**Additional context**
 --num-layers 60 \
    --hidden-size 5120 \
    --ffn-hidden-size 12288 \
    --num-attention-heads 128 \
    --seq-length 4096 \
    --max-position-embeddings 4096 \
    --multi-latent-attention \
    --kv-lora-rank 512 \
    --q-lora-rank 1536 \
    --moe-ffn-hidden-size 1536 \
    --moe-router-topk 6 \
    --num-experts 160 \
    --decoder-first-pipeline-num-layers 2 \
    --decoder-last-pipeline-num-layers 2 \
    --moe-aux-loss-coeff 1e-2 \
    --moe-shared-expert-intermediate-size  3072 \
    --moe-layer-freq [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1] \
    --expert-model-parallel-size 8 \
    --moe-grouped-gemm \
    --moe-router-load-balancing-type aux_loss \
    --qk-head-dim 128 \
    --qk-pos-emb-head-dim 64 \
    --v-head-dim 128 \
    --rotary-base 10000 \
    --no-rope-fusion \
    --no-bias-dropout-fusion \
    --no-persist-layer-norm \
    --micro-batch-size 1 \

**Comment by Desperadoze** [comment]
![Image](https://github.com/user-attachments/assets/116cc3fe-2ff2-418c-8d5e-441ae6f8cd76)  The lower two curves are GQA and MHA, and the upper two curves are MLA.

**Comment by yanring** [comment]
Thank you for reporting the issue. @xxuwenc @Shunkangz pls help take a look.
Are you using the same hyper-parameters, only changing the attention type for these experiments? and could you also share the grad-norm curve?

**Comment by Desperadoze** [comment]
The difference between these four curves: In the two using MLA, black has a capacity of 4, green is dropless, orange is GQA, I forgot if yellow is GQA with more grouping or MHA, all other configuration items are exactly the same. The grade norm of MLA will increase to 1000-2000 after 4000 steps

![Image](https://github.com/user-attachments/assets/5555d44e-e0af-4d6c-9932-f13d8318c0af)







> Thank you for reporting the issue. [@xxuwenc](https://github.com/xxuwenc) [@Shunkangz](https://github.com/Shunkangz) pls help take a look. Are you using the same hyper-parameters, only changing the attention type for these experiments? and could you also share the grad-norm curve?

**Comment by yanring** [comment]
From what I saw on the grad_norm curve, the orange one increase to between 1000 and 2000 after 4000 steps. Is the orange curve MLA or MHA?

**Comment by Desperadoze** [comment]
> From what I saw on the grad_norm curve, the orange one increase to between 1000 and 2000 after 4000 steps. Is the orange curve MLA or MHA?

After these four curves, I ran some experiments again. The screenshots were taken at different times, so the colors of the curves in the two graphs did not correspond. The above description is for the loss graph. The purple and orange colors in the following graph correspond to the green and black colors in the loss graph. The grad norm of GQA and MHA has always been within the normal range. Due to the scale, it looks like two overlapping bottom straight lines in the grad norm graph

**Comment by yanring** [comment]
I see, thanks! We will run some experiments on v0.11(the latest stable branch) first to see if we can reproduce your issue and will get back to you.
By the way, it would be super helpful if you could share the scripts and the W&B report if available.

**Comment by Desperadoze** [comment]
This is my running configuration item. Compared with the actual training script, the data related and path related parts have been deleted. I am sure it is not  caused by the data, because after replacing  different data, the loss curve of MLA is still abnormal. The megatron-lm version used is the code in the image nvcr. io/nvidia/nemo: 25.02.rc1，Thanks  for your patience and help！

[deepseek_236.txt](https://github.com/user-attachments/files/19065318/deepseek_236.txt)

**Comment by Victarry** [comment]
@Desperadoze  Thank you for reporting this issue and sharing detailed configuration details.

We tried to reproduce the divergence issue in [MCore 0.11](https://github.com/NVIDIA/Megatron-LM/tree/core_r0.11.0) with a small model to see if it's caused by MLA implementation error, and the loss seems normal.

![Image](https://github.com/user-attachments/assets/7379f1f8-b800-4f96-9760-f98ff000c089)

We noticed that your provided config doesn't enable qk layernorm by `--qk-layernorm`, this may cause training instability as said in DeepSeek-V2 paper. 

<img width="1425" alt="Image" src="https://github.com/user-attachments/assets/1e13024e-a5e5-4be7-bce3-42e47b8884e1" />

And the above divergence looks similar to the training instability without qk-layernorm in https://arxiv.org/pdf/2309.14322.

Could you have a try to enable `--qk-layernorm` and see if the divergence issue still exists?

**Comment by Desperadoze** [comment]
Thank you very much for your suggestions and time and effort. After adding --qk-layernorm item and using core0.11, everything looks normal. I will pay attention to the downstream task scores of the model implemented with megatron core's MLA to ensure that the entire functionality is correct .

**Comment by Victarry** [comment]
Great! Thanks for your feedback.

**Comment by yeontaek** [comment]
I have another question, but I'm curious about how well the performance turns out with this setup.

**Comment by SeunghyunSEO** [comment]
> **Describe the bug** when using mla by megatron-core 0.10.0,loss was unexpected
> 
> **To Reproduce** use /nvcr.io/nvidia/nemo:25.02.rc1 image start traing
> 
> **Additional context** Add any other context about the problem here. --num-layers 60 --hidden-size 5120 --ffn-hidden-size 12288 --num-attention-heads 128 --seq-length 4096 --max-position-embeddings 4096 --multi-latent-attention --kv-lora-rank 512 --q-lora-rank 1536 --moe-ffn-hidden-size 1536 --moe-router-topk 6 --num-experts 160 --decoder-first-pipeline-num-layers 2 --decoder-last-pipeline-num-layers 2 --moe-aux-loss-coeff 1e-2 --moe-shared-expert-intermediate-size 3072 --moe-layer-freq [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1] --expert-model-parallel-size 8 --moe-grouped-gemm --moe-router-load-balancing-type aux_loss --qk-head-dim 128 --qk-pos-emb-head-dim 64 --v-head-dim 128 --rotary-base 10000 --no-rope-fusion --no-bias-dropout-fusion --no-persist-layer-norm --micro-batch-size 1 \

may I ask about other parallelism setups? I keep failing to reproduce DeepSeek V2.
MLA causes high memory pressure, so I manage it with selective checkpointing but it leads to an OOM error after a few hundred steps, once load balancing stabilizes.
if I'm right, theoretically one device can consume (4096x8/160x(6+2))x(160/8)=32k (considering shared experts) when mbsz=1, seqlen=4096 (typical dsv2 setup i guess) and EP=8 and token drop is used. 
i guess this large batch cause OOM.
p.s.) while it's ffn_dim is 1/8 times smaller than dense, 32k * moe_ffn_dim is same as dense's, but it requires larger activation memory because it's input hidden dim is same as dense's

**Comment by Desperadoze** [comment]
> > **Describe the bug** when using mla by megatron-core 0.10.0,loss was unexpected
> > **To Reproduce** use /nvcr.io/nvidia/nemo:25.02.rc1 image start traing
> > **Additional context** Add any other context about the problem here. --num-layers 60 --hidden-size 5120 --ffn-hidden-size 12288 --num-attention-heads 128 --seq-length 4096 --max-position-embeddings 4096 --multi-latent-attention --kv-lora-rank 512 --q-lora-rank 1536 --moe-ffn-hidden-size 1536 --moe-router-topk 6 --num-experts 160 --decoder-first-pipeline-num-layers 2 --decoder-last-pipeline-num-layers 2 --moe-aux-loss-coeff 1e-2 --moe-shared-expert-intermediate-size 3072 --moe-layer-freq [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1] --expert-model-parallel-size 8 --moe-grouped-gemm --moe-router-load-balancing-type aux_loss --qk-head-dim 128 --qk-pos-emb-head-dim 64 --v-head-dim 128 --rotary-base 10000 --no-rope-fusion --no-bias-dropout-fusion --no-persist-layer-norm --micro-batch-size 1 \
> 
> may I ask about other parallelism setups? I keep failing to reproduce DeepSeek V2. MLA causes high memory pressure, so I manage it with selective checkpointing but it leads to an OOM error after a few hundred steps, once load balancing stabilizes. if I'm right, theoretically one device can consume (4096x8/160x(6+2))x(160/8)=32k (considering shared experts) when mbsz=1, seqlen=4096 (typical dsv2 setup i guess) and EP=8 and token drop is used. i guess this large batch cause OOM. p.s.) while it's ffn_dim is 1/8 times smaller than dense, 32k * moe_ffn_dim is same as dense's, but it requires larger activation memory because it's input hidden dim is same as dense's

The configuration in the issue above is the parallel configuration I am using. Can you try changing the parameters related to recompute? Under the above configuration, the memory usage is approximately between 64-68G, which is a relatively safe value

**Comment by SeunghyunSEO** [comment]
> > > **Describe the bug** when using mla by megatron-core 0.10.0,loss was unexpected
> > > **To Reproduce** use /nvcr.io/nvidia/nemo:25.02.rc1 image start traing
> > > **Additional context** Add any other context about the problem here. --num-layers 60 --hidden-size 5120 --ffn-hidden-size 12288 --num-attention-heads 128 --seq-length 4096 --max-position-embeddings 4096 --multi-latent-attention --kv-lora-rank 512 --q-lora-rank 1536 --moe-ffn-hidden-size 1536 --moe-router-topk 6 --num-experts 160 --decoder-first-pipeline-num-layers 2 --decoder-last-pipeline-num-layers 2 --moe-aux-loss-coeff 1e-2 --moe-shared-expert-intermediate-size 3072 --moe-layer-freq [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1] --expert-model-parallel-size 8 --moe-grouped-gemm --moe-router-load-balancing-type aux_loss --qk-head-dim 128 --qk-pos-emb-head-dim 64 --v-head-dim 128 --rotary-base 10000 --no-rope-fusion --no-bias-dropout-fusion --no-persist-layer-norm --micro-batch-size 1 \
> > 
> > may I ask about other parallelism setups? I keep failing to reproduce DeepSeek V2. MLA causes high memory pressure, so I manage it with selective checkpointing but it leads to an OOM error after a few hundred steps, once load balancing stabilizes. if I'm right, theoretically one device can consume (4096x8/160x(6+2))x(160/8)=32k (considering shared experts) when mbsz=1, seqlen=4096 (typical dsv2 setup i guess) and EP=8 and token drop is used. i guess this large batch cause OOM. p.s.) while it's ffn_dim is 1/8 times smaller than dense, 32k * moe_ffn_dim is same as dense's, but it requires larger activation memory because it's input hidden dim is same as dense's
> 
> The configuration in the issue above is the parallel configuration I am using. Can you try changing the parameters related to recompute? Under the above configuration, the memory usage is approximately between 64-68G, which is a relatively safe value

oh, did you use recompute method? for all transformer blocks?

**Comment by Desperadoze** [comment]
> > > > **Describe the bug** when using mla by megatron-core 0.10.0,loss was unexpected
> > > > **To Reproduce** use /nvcr.io/nvidia/nemo:25.02.rc1 image start traing
> > > > **Additional context** Add any other context about the problem here. --num-layers 60 --hidden-size 5120 --ffn-hidden-size 12288 --num-attention-heads 128 --seq-length 4096 --max-position-embeddings 4096 --multi-latent-attention --kv-lora-rank 512 --q-lora-rank 1536 --moe-ffn-hidden-size 1536 --moe-router-topk 6 --num-experts 160 --decoder-first-pipeline-num-layers 2 --decoder-last-pipeline-num-layers 2 --moe-aux-loss-coeff 1e-2 --moe-shared-expert-intermediate-size 3072 --moe-layer-freq [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1] --expert-model-parallel-size 8 --moe-grouped-gemm --moe-router-load-balancing-type aux_loss --qk-head-dim 128 --qk-pos-emb-head-dim 64 --v-head-dim 128 --rotary-base 10000 --no-rope-fusion --no-bias-dropout-fusion --no-persist-layer-norm --micro-batch-size 1 \
> > > 
> > > 
> > > may I ask about other parallelism setups? I keep failing to reproduce DeepSeek V2. MLA causes high memory pressure, so I manage it with selective checkpointing but it leads to an OOM error after a few hundred steps, once load balancing stabilizes. if I'm right, theoretically one device can consume (4096x8/160x(6+2))x(160/8)=32k (considering shared experts) when mbsz=1, seqlen=4096 (typical dsv2 setup i guess) and EP=8 and token drop is used. i guess this large batch cause OOM. p.s.) while it's ffn_dim is 1/8 times smaller than dense, 32k * moe_ffn_dim is same as dense's, but it requires larger activation memory because it's input hidden dim is same as dense's
> > 
> > 
> > The configuration in the issue above is the parallel configuration I am using. Can you try changing the parameters related to recompute? Under the above configuration, the memory usage is approximately between 64-68G, which is a relatively safe value
> 
> oh, did you use recompute method? for all transformer blocks?

    --recompute-granularity full \
    --recompute-num-layers 1 \
    --recompute-method uniform \
